cmd/snap-update-ns/mount-entry-change-test.c

```diff
@@ -0,0 +1,334 @@
+	desired = sc_load_mount_profile("desired.fstab");
+	g_test_queue_destroy((GDestroyNotify)
+			     sc_free_mount_entry_list, desired);
+
+	struct sc_mount_change *change;
+	change = sc_compute_required_mount_changes(desired, current);
+	g_test_queue_destroy((GDestroyNotify)
+			     sc_mount_change_free_chain, change);
+	test_assert_change_list(change,
+				// mount 2nd entry
+				&test_entry_2, SC_ACTION_MOUNT, NULL);
+}
+
+// Scenario: the current profile contains one entry and the desired profile
+// contains one entry but they are different. We should see the unmount
+// followed by the mount.
+static void test_sc_compute_required_mount_changes__scenario4()
+{
+	struct sc_mount_entry_list *current;
+	sc_test_write_lines("current.fstab", test_entry_str_1, NULL);
+	current = sc_load_mount_profile("current.fstab");
+	g_test_queue_destroy((GDestroyNotify)
+			     sc_free_mount_entry_list, current);
+
+	struct sc_mount_entry_list *desired;
+	sc_test_write_lines("desired.fstab", test_entry_str_2, NULL);
+	desired = sc_load_mount_profile("desired.fstab");
+	g_test_queue_destroy((GDestroyNotify)
+			     sc_free_mount_entry_list, desired);
+
+	struct sc_mount_change *change;
+	change = sc_compute_required_mount_changes(desired, current);
+	g_test_queue_destroy((GDestroyNotify)
+			     sc_mount_change_free_chain, change);
+	test_assert_change_list(change, &test_entry_1,
+				SC_ACTION_UNMOUNT,
+				&test_entry_2, SC_ACTION_MOUNT, NULL);
+}
+
+// Scenario: desired A, B current B, C behaves correctly (B is untouched).
+static void test_sc_compute_required_mount_changes__scenario5()
+{
+	struct sc_mount_entry_list *current;
+	sc_test_write_lines("current.fstab",
+			    "B B B B 0 0", "C C C C 0 0", NULL);
+	current = sc_load_mount_profile("current.fstab");
+	g_test_queue_destroy((GDestroyNotify)
+			     sc_free_mount_entry_list, current);
+
+	struct sc_mount_entry_list *desired;
+	sc_test_write_lines("desired.fstab",
+			    "A A A A 0 0", "B B B B 0 0", NULL);
+	desired = sc_load_mount_profile("desired.fstab");
+	g_test_queue_destroy((GDestroyNotify)
+			     sc_free_mount_entry_list, desired);
+
+	struct sc_mount_change *change;
+	change = sc_compute_required_mount_changes(desired, current);
+	g_test_queue_destroy((GDestroyNotify)
+			     sc_mount_change_free_chain, change);
+	const struct sc_mount_entry C = {
+		.entry = {
+			  "C", "C", "C", "C"}
+	};
+	const struct sc_mount_entry A = {
+		.entry = {
+			  "A", "A", "A", "A"}
+	};
+	test_assert_change_list(change,
+				&C, SC_ACTION_UNMOUNT,
+				&A, SC_ACTION_MOUNT, NULL);
+}
+
+// Scenario: desired A, A/B, current: A A/B with the tweak that A changes
+// subtly (e.g. different type of mount vs what we had earlier).
+static void test_sc_compute_required_mount_changes__scenario6()
+{
+	struct sc_mount_entry_list *current;
+	sc_test_write_lines("current.fstab",
+			    "/dev/sda1 /foo ext4 rw 0 0",
+			    "/dev/loop7 /foo/bar squashfs ro 0 0", NULL);
```

Comment:
Not sure what the structure of the code is when it comes to sorting - but if sc_load_mount_profile() does the sorting (which is crucial for the algorithm) then I think we should write out the lines above unsorted so that we verify that this is taken into account.